add react use const (#12)
navarroaxel/navarroaxel.github.io: blog/you-cant-create-a-constant-using-usememo-in-react.md

# You can't create a constant using useMemo in React

> There are 3 possible options to create a constant in React: `useState`, `useMemo` and `useRef`. We're going to see which one is the best way to create real constants (not constant-like values).

Reviewer: ```suggestion
There are 3 possible options to create a real constant (not constant-like values) in React: `useState`, `useMemo` and `useRef`. Now we're going to see which one is the best.
```

Reviewer: This paragraph could be rewritten in a diff order, but I think you should not repeat sentences that way

> We have 3 possible options to create a constant in React: `useState`, `useMemo` and `useRef`. We're going to review which is the best way to create real constants (not constant-like values) in React.

Reviewer 2: ```suggestion
There are 3 possible options to create a constant in React: `useState`, `useMemo` and `useRef`. We're going to see which one is the best way to create real constants (not constant-like values).
```

> I talked about the [`useConst`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/src/use-const.ts) hook from [chakra-ui](https://chakra-ui.com) in a [previous post](https://dev.to/cloudx/chakra-ui-the-hidden-treasures-1gl9), and now, I found the same hook in [Fluent UI](https://www.npmjs.com/package/@fluentui/react-hooks#useconst), we are going to review why they've implemented it.

Reviewer: ```suggestion
I talked about the [`useConst`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/src/use-const.ts) hook from [chakra-ui](https://chakra-ui.com) in a [previous post](https://dev.to/cloudx/chakra-ui-the-hidden-treasures-1gl9), and I recently found the same hook in [Fluent UI](https://www.npmjs.com/package/@fluentui/react-hooks#useconst) so we are going to see why they've implemented it.
```

> I mention in a [previous post](https://dev.to/cloudx/chakra-ui-the-hidden-treasures-1gl9) about the [`useConst`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/src/use-const.ts) hook from [chakra-ui](https://chakra-ui.com), and now, I found the same hook in [Fluent UI](https://www.npmjs.com/package/@fluentui/react-hooks#useconst), we are going to review why these hooks implementation.

Reviewer 2: ```suggestion
I talked about the [`useConst`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/src/use-const.ts) hook from [chakra-ui](https://chakra-ui.com) in a [previous post](https://dev.to/cloudx/chakra-ui-the-hidden-treasures-1gl9), and now, I found the same hook in [Fluent UI](https://www.npmjs.com/package/@fluentui/react-hooks#useconst), we are going to review why they've implemented it.
```

> I talked about the [`useConst`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/src/use-const.ts) hook from [chakra-ui](https://chakra-ui.com/docs/hooks/use-const) in a [previous post](https://dev.to/cloudx/chakra-ui-the-hidden-treasures-1gl9), and I recently found the same hook in [Fluent UI](https://www.npmjs.com/package/@fluentui/react-hooks#useconst) so we are going to see why they've implemented it.

Reviewer 3: ```suggestion
I talked about the [`useConst`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/src/use-const.ts) hook from [chakra-ui](https://chakra-ui.com/docs/hooks/use-const) in a [previous post](https://dev.to/cloudx/chakra-ui-the-hidden-treasures-1gl9), and I recently found the same hook in [Fluent UI](https://www.npmjs.com/package/@fluentui/react-hooks#useconst) so we are going to see why they implemented it.
```

## Remember useMemo

> The first thought that comes to mind when we look for a constant in a render function is to use the [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) hook. This is a good option, but only if your "constant" value shouldn't change based on the dependencies.

Reviewer: ```suggestion
The first thing I think of when I look for a constant in a render function is the use of [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) hook. This is a good option, but only if your "constant" value shouldn't change based on the dependencies.
```

> The first think that comes to our mind when we look for a constant in a render function is to use the [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) hook. This is a good option, but only if your "constant" value should ever change based on dependencies.

Reviewer 2: ```suggestion
The first thought that comes to mind when we look for a constant in a render function is to use the [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) hook. This is a good option, but only if your "constant" value shouldn't change based on the dependencies.
```

```javascript
const memoizedValue = useMemo(
  () => computeExpensiveValue(a, b),
  [a, b]
);
```

> There is no warranty that the returned value of `useMemo` is always the same even when the dependencies didn't change. React can run the factory function if the tool believes that the value should be re created.

Reviewer: ```suggestion
There is no guarantee that the returned value of `useMemo` is always the same even if the dependencies didn't change. React can run the factory function if the tool believes that the value should be re created.
```

Reviewer: Could we add a connector between these sentences?

Reviewer 2: ```suggestion
There is no guarantee that the returned value of `useMemo` is always the same even when the dependencies didn't change. For this reason React runs the factory function if the tool believes that the value should be re created.
```

🧠 The official documentation says:

> Write your code so that it still works without `useMemo` — and then add it to optimize performance.

## The useState option

> If a state didn't change, then it's a constant.

Reviewer: ```suggestion
If a state doesn't change, then it's a constant.
```

```javascript
const [value] = useState(initialValue);
```

> Yeah, we can create a constant with [`useState`](https://reactjs.org/docs/hooks-reference.html#usestate) but it's expensive due to reducer handling which we don't need.

Reviewer: ```suggestion
Yeah, we can create constants using [`useState`](https://reactjs.org/docs/hooks-reference.html#usestate), but it's expensive due to the reducer handling elements we don't need.
```

Reviewer: `elements` can be replaced with a word of your preference

🧙‍♂️ If you want to learn how React handles its hooks inside the Fiber reconciler, you should read the [Fiber hooks](https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactFiberHooks.new.js) code.

## A useRef approach

> The [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) hook can hold a value, and doesn't have an internal reducer or checks on the dependencies array, also React doesn't re-create the value for performance reasons.

Reviewer: ```suggestion
The [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) hook can hold a value, and it neither has an internal reducer nor checks on the dependencies array. Also, React doesn't re-create the value for performance reasons.
```

> The [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) hook can hold a value, and doesn't have an internal reducer or checks the dependencies array, also React doesn't re-create the value for performance reasons.

Reviewer 2: ```suggestion
The [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) hook can hold a value, and doesn't have an internal reducer or checks on the dependencies array, also React doesn't re-create the value for performance reasons.
```

```javascript
const useConst = (initialValue) => {
  const ref = React.useRef();
  if (ref.current === undefined) {

>     ref.current = typeof initialValue === 'function' ? initialValue() : initialValue,

Reviewer: Is the `,` right at the end of this line?

Author: ```suggestion
    ref.current = typeof initialValue === 'function' ? initialValue() : initialValue;
```

  }
  return ref.current;
}
```

💡 This is the cheapest way to create a constant value over the lifecycle of a component.

## Conclusion

> ❌ Don't use `useState` to create constants because it's expensive.

Reviewer 3: he-man meme, and remember kids... don't use useState because it's expensive

Reviewer 2: ![image](https://user-images.githubusercontent.com/41996135/146237658-40e42b3a-74e2-4fee-8d9e-a0f2751b791d.png)

> 🥸 If you need a constant-like value based on some dependencies `useMemo` is for you, but your code should still work without this hook.

Reviewer: ```suggestion
🥸 If you need a constant-like value based on some dependencies then `useMemo` is for you (but your code should still work without this hook).
```

> ✅ If you need a real constant over the lifecycle of a component, `useRef` is the solution that you need, and some UI libraries like Chakra UI or Fluent UI provides a built-in `useConst` for this.

Reviewer: ```suggestion
✅ If you need a real constant over the lifecycle of a component, `useRef` is the solution that you need, and remember that some libraries like Chakra UI or Fluent UI provides a built-in `useConst` for this.
```
